### DISCUSSION TITLE
Search/Multi-Database Extensions

### DISCUSSION THREAD

**Discussion by gdj0nes** [sqlalchemy/sqlalchemy#8616] [Ideas]
I work on at a vector database company called Pinecone. Many of users combine our database with a second transactional database like Postgres or MySQL to entities. Managing data across two systems is a frequent challenge and we see many near identical implementations. 

I came across [Hibernate Search](https://hibernate.org/search/) and wondered if it straightforward to extend SQLAlchemy in a similar way. Ideally we would create an extension that allowed users to keep data in sync between a vector index while exposing a nice experience for search and fetching entities.

**Comment by zzzeek** [comment]
just took a look at hibernate search, and while I haven't done programming with Hibernate in nearly 20 years, it does not *seem* to be too much "integrated" into hibernate as much as it is it's "similar" to Hibernate in how it does configuration.   The querying interface it provides seems to be [it's own thing](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl) (yes it's within the "hibernate" namespace, but is not the SQL / ORM that Hibernate is famous for), at least very much not Hibernate's traditional API and not [HQL](https://docs.jboss.org/hibernate/core/3.3/reference/en-US/html/queryhql.html), which is more comparable to SQLAlchemy's DSL that is very much 1-1 with SQL.  

So I dont know that there's much of an "integration" with non-relational search engines that don't have a SQL interpreter, or something extremely similar to SQL.   There have been integrations of SQLAlchemy with non-SQL datasources, if I look over [here](https://docs.sqlalchemy.org/en/14/dialects/index.html) I see there's one for [elasticsearch](https://github.com/preset-io/elasticsearch-dbapi/), but that seems to take advantage of a SQL interface for elastic search.

from my end, if your search engine has a SQL interface, then it's a straight path, if not, then it's not so clear.   SQLAlchemy doesnt actually get super high marks for being "easy" in most cases :)  and it's very welded to how SQL works, so if you are querying some other kind of structure, you'd be making something different, like Hibernate's thing above is not really very hibernate-like, it's using a separate search DSL.

I think the best thing to start with is, you have your search engine, what kind of Python API do you want?   Just design it and come up with how it looks and what the use cases are.   If you familiarize with SQLAlchemy ahead of time, you can see what overlap might be worth taking advantage of.   Then the question of, "do we try to extend SQLAlchemy, or just make our own thing that looks very similar" can be approached.  if that makes sense.

**Comment by gdj0nes** [comment]
Thanks for the reply, I really appreciate it and have used SQLAlchemy for a number of projects previously!

We do not have a SQL like interface for the index and I see how that it would be more challenging to extend SQLAlchemy after digging through the code in more detail. 

I found the Hibernate Search approach interesting because it allowed users to define the relational and non-relational pieces in one place. The decorator interface in particular i found appealing.

```java
@Entity
// This entity is mapped to an index
@Indexed
public class Book {

    // The entity ID is the document ID
    @Id
    @GeneratedValue
    private Integer id;

    // This property is mapped to a document field
    @FullTextField
    private String title;

    @ManyToMany
    // Authors will be embedded in Book documents
    @IndexedEmbedded
    private Set<Author> authors = new HashSet<>();
}
```

And that persistence to both data stores in a single method

```java 
Author author = new Author();
author.setName( "Isaac Asimov" );

Book book = new Book();
book.setTitle( "The Caves Of Steel" );
book.getAuthors().add( author );
author.getBooks().add( book );

entityManager.persist( author );
entityManager.persist( book );
```

I've looked a bit through the code of the Hive extension but it doesn't seem like there is a way of defining two engines at once.

**Reply by zzzeek** [reply]
so you could use SQLAlchemy ORM mappings as they are, in order to build a non-relational search API.   you could have the usual MetaData/Table/Mapper concepts in use, then you just iterate through the properties, like using [mapper.attrs](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=mapper+attrs#sqlalchemy.orm.Mapper.attrs), to know what the fields are.    The API would be ignoring a large portion of SQLAlchemy Core /ORM like foreign keys and relationships and all that but you could repurpose it in this way.    It might be seen as a little bit hacky though.  

Also, I would make sure to look at how SQLAlchemy 2.0 is doing these things now.  We have an [all new declarative system](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#orm-declarative-models) that is deeply inspired by Python dataclasses and also includes new integrations with dataclasses.    

For search API, it might be very feasible to simply use [dataclasses](https://docs.python.org/3/library/dataclasses.html) directly to define the "models".    Or even just straight Python annotated classes.    Using the [pep-593 Annotated](https://peps.python.org/pep-0593/) objects, you could encapsulate all the additional arguments within Annotated structures.   [SQLAlchemy is pushing in that direction](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapping-whole-column-declarations-to-python-types) in the 2.0 series.

Basically I'd look to the latest Python features above for inspiration and not look too closely at Hibernate / Java stuff, which works from a very different perspective.   SQLAlchemy has a lot of Hibernate-isms in it, which I did intentionally, and while it allowed us to have some nice features up front without me having to think of them, it's also put a little bit of an overly "enterprisey" feel to our APIs that people struggle with quite often.  for a popular library you want to be extremely careful about using anything J2EE as inspiration, basically.
